File: mpi/exercises/exercise_2/solution.c
#include <stdio.h>
#include <mpi.h>

/**
 * @brief Solution to the simple send-receive in MPI.
 **/
int main(int argc, char* argv[])
{
	// 1) Tell MPI to start
	MPI_Init(&argc, &argv);

	// 2) Check that the application is run with 2 MPI processes
	int comm_size;
	MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
	if(comm_size != 2)
	{
		printf("This application must be run with 2 MPI processes, not %d.\n", comm_size);
		MPI_Abort(MPI_COMM_WORLD, -1);
	}

	// 3) Get my rank
	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

	// 4) If my rank is 0, I am the sender
	if(my_rank == 0)
	{
		// 4.1) Print the value to send
		int value_to_send = 12345;
		printf("I am process %d and I send value %d.\n", my_rank, value_to_send);

		// 4.2) Send the value
		MPI_Send(&value_to_send, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
	}	
	// 5) If my rank is 1, I am the receiver
	else
	{
		// 5.1) Receive the value
		int value_received = 0;
		MPI_Recv(&value_received, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

		// 5.2) Print the value received
		printf("I am process %d and I received value %d.\n", my_rank, value_received);
	}

	// 6) Tell MPI to stop
	MPI_Finalize();

	return 0;
}

File: mpi/exercises/exercise_2/solution.f90
!> @brief Solution to the simple send-receive in MPI.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: comm_size
    INTEGER :: my_rank
    INTEGER :: value

    ! 1) Tell MPI to start
    CALL MPI_Init(ierror)

    ! 2) Check that the application is run with 2 MPI processes
    CALL MPI_Comm_size(MPI_COMM_WORLD, comm_size, ierror)
    IF (comm_size .NE. 2) THEN
        WRITE(*, '(A,I0,A)') 'This application must be run with 2 MPI processes, not ', comm_size, '.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierror)
    END IF

    ! 3) Get my rank
    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)

    ! 4) If my rank is 0, I am the sender
    IF (my_rank .EQ. 0) THEN
        ! 4.1) Print the value to send
        value = 12345
        WRITE(*, '(A,I0,A,I0,A)') 'I am process ', my_rank, ' and I send value ', value, '.'

        ! 4.2) Send the value
        CALL MPI_Send(value, 1, MPI_INTEGER, 1, 0, MPI_COMM_WORLD, ierror)    
    ! 5) If my rank is 1, I am the receiver
    ELSE
        ! 5.1) Receive the value
        CALL MPI_Recv(value, 1, MPI_INTEGER, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror)

        ! 5.2) Print the value received
        WRITE(*, '(A,I0,A,I0,A)') 'I am process ', my_rank, ' and I received value ', value, '.'
    END IF

    ! 6) Tell MPI to stop
    CALL MPI_Finalize(ierror)
END PROGRAM main